### ISSUE TITLE
Absolute timelocks support

### ISSUE THREAD

**Issue by darosior** [revault/minisafe#48]
For now, we restrict the descriptors to relative timelocks. This presents two limitations:
- The wallet has to be used in order to "rotate" the coins (re-start the timelock)
- The timelock value is 2 bytes and thus can't be greater than `2**16` (65536), that is approximately 15 months:
   ```python
   >>> 2**16 / 6 / 24 / 365
   1.2468797564687975
   ```

This makes it not a very good fit for cold wallets that aren't accessed, and for which all coins are rotated, at least around once per year.
This also makes it hard for a non-technical heir to sweep the coins of a deceased parent that was using Minisafe: not all coins may be available at the same height.

The main and obvious drawback of using an absolute timelock is that the descriptor would be have an EOL. After (or shortly before) the expiration of the absolute timelock, one would have to sweep their coins to a new descriptor with a later expiration date. Still, this tradeoff may be worth it for some usages such as the one depicted above (longer timelocks without the need to rotate, simpler inheritance recovery).

Thoughts?

**Comment by edouardparis** [comment]
Good points ! One Absolute timelock is easier to remember for a user using the software as cold wallet support.

**Comment by pythcoiner** [comment]
Absolute timelock can be also usefull for birthday unlock: i send sats to my kids saving wallet but allow him to spend only after age of 18.

**Comment by darosior** [comment]
There's been some interest in using absolute timelocks with Liana recently. Unfortunately it would require some pretty involved changes, especially to the user interface.

**Comment by pythcoiner** [comment]
> There's been some interest in using absolute timelocks with Liana recently. Unfortunately it would require some pretty involved changes, especially to the user interface.

what about first allow user to import descriptor with absolute timelock?

**Comment by pythcoiner** [comment]
> Well.. The software needs to work with those descriptors before we let users import them.

i was guessing that the main UI changes needed were about policy creation....

**Comment by nondiremanuel** [comment]
> Thanks for linking to this issue. As the original post describes, with absolute timelocks one would not need to rotate every <15 months.
> 
> I think it should not be much work to make lianad be able to parse descriptors which have older(n), with n<65k, replaced by after(n) with n larger. The descriptor itself could be changed manually in the json backup file (and using Bitcoin core for the changed fingerprint), so it would not be necessary to change the wallet creation GUI. The GUI would only need to show that the coins are movable after blockheight n instead of after n blocks.

Unfortunately, the main effort, as I was mentioning in the other issue, would be related to the user experience since at timelock refresh we should rotate all coins to a new descriptor. This will probably be easier in the future when we will have the key rotation feature.
By the way, I'm interested in the rationale behind your request in order to better guide our development and priority: do you think the 15 months is too low? Are you not afraid of the funds to be locked for too much time in case you use a long timelock duration? Do you have any specific detail on the use case you had in mind? Thanks!

**Comment by chuber11** [comment]
yes, in my opinion 15 months is too low, since one would need rotate every < 15 month, which could become costly in the future.

The usecase would be to have a backup which would be accessible after e.g. 4-10 years. I don't think such a locking time is too long, since it's only for backup. To get the same functionality currently, one would need to presign a transaction with timelock every time a transaction is received.

Also, I don't think rotating to a new descriptor is a big problem. It would be sending from one wallet with an absolute timelock which ends too soon to a wallet with the same keys and a later absolute timelock.

**Comment by pythcoiner** [comment]
In fact there is lot of underlying changes to be done before introducing such features:
- The descriptor analysis logic must be updated.
- The wallet logic must be able to watch several descriptors, as every time you create a new

There is also some painfull limitations due to how signing devices actually handle descriptors, for instance you'll not be able to spend 2 inputs from different descriptor in a safe way...

Some points of the design must be clarified:
- Every time a new timelock value is used, the used must have a sane way to backup the newly generated descriptor (preferably in an automatic way).
- Every time a new timelock value is used, the user MUST register/validate the newly created descriptor on every signing device 
- Arbitrary timelocks values should not be accepted, we should go to a scheme where valid timelocks values follow and incremental scheme, like a value must be a modulo of a time-unit of 4000 blocks (nearly one month) for instance, in order to make it easily bruteforceable if some intermediary backup is lost.
- Some metadata not contained in the descriptor itself will be needed to be backed-up also, like the previously mentioned time-unit.
- Interoperability between others software wallet can be hard to achieve because possibly different technical choices...

Note: I've just mention here what I have in mind by just giving some thought, it's very likely I'm overlooking many important points, but my general POV on this topic is that it's absolutely not trivial, and I don't think it worth allocating much design/dev time unless there is a STRONG demand for this.
